### PR TITLE
Fix warning: 'creating a shared reference to mutable static is discouraged'

### PR DESCRIPTION
**Pull request by bugadani**

### CODE REVIEW

**Review by MabezDev — COMMENTED** [review]
Any reason we don't just return a copy of `Clocks`?

**Comment by bugadani** [comment]
It's relatively big 😅 but we could

**Review by MabezDev — APPROVED** [review]
> It's relatively big 😅 but we could

Fair enough. If a copy is required at the call site a user can always deref.

File: esp-hal/src/clock/mod.rs [hunk 1]
```diff
@@ -319,15 +319,16 @@ impl Clocks {
         })
     }
 
-    fn try_get() -> Option<&'static Clocks> {
+    fn try_get<'a>() -> Option<&'a Clocks> {
```
[the remark this answers is not in the document]
Comment (bugadani) [reply esp-hal/src/clock/mod.rs:322]: Yes and no. This doesn't prevent you from taking a `'static` reference, but also doesn't force you to do that either. Shouldn't make a difference right now.

**Review by Dominaezzz — APPROVED** [review]
A bit weird/incorrect that `init` uses a critical section and `get` doesn't, but that existed before this PR.

PR LGTM as is

File: esp-hal/src/clock/mod.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (Dominaezzz) [reply esp-hal/src/clock/mod.rs:322]: Fair

**Comment by MabezDev** [comment]
> A bit weird/incorrect that init uses a critical section and get doesn't, but that existed before this PR.

Currently (and probably forever?) `init` can only be called once and once clocks are configured that's it. Meaning this sort of behaves like `OnceCell`.

**Comment by Dominaezzz** [comment]
> > A bit weird/incorrect that init uses a critical section and get doesn't, but that existed before this PR.
> 
> Currently (and probably forever?) `init` can only be called once and once clocks are configured that's it. Meaning this sort of behaves like `OnceCell`.

Sure but if it's only called once, what the critical section for? :smile:

**Comment by bugadani** [comment]
The critical section is there to make sure that if you manage to somehow call `Clocks::get` from an ISR before we `init`ed, you don't get a half-initialized mess. You either get a panic because you're doing something bad, or you get the inited struct, nothing in-between.

Embassy actually manages to call `try_get()` before `init`, which is extra fun. Thankfully it's only in the thread context.

**Comment by Dominaezzz** [comment]
> The critical section is there to make sure that if you manage to somehow call `Clocks::get` from an ISR before we `init`ed, you don't get a half-initialized mess.

That's what I figured but `Clocks::get` doesn't have a critical section right now

**Comment by Dominaezzz** [comment]
Oh I think I get what you're saying, the critical section in `init` prevents any other context from calling `get`? I suppose this would only apply to single core chips.

**Comment by bugadani** [comment]
I don't think you can enable the second core without unsafe code right now (you need a peripheral for it that only `init` returns safely), and there's no reason to tank performance just because we expect someone to commit crimes.

**Comment by Dominaezzz** [comment]
> no reason to tank performance just because we expect someone to commit crimes.

Fair enough :smile:

**Pull request merged**